### PR TITLE
[MLlib] SPARK-5954: Top by key

### PR DESCRIPTION
**Pull request by coderxiang**
This PR implements two functions
- `topByKey(num: Int): RDD[(K, Array[V])]` finds the top-k values for each key in a pair RDD. This can be used, e.g., in computing top recommendations.
- `takeOrderedByKey(num: Int): RDD[(K, Array[V])]` does the opposite of `topByKey`

The `sorted` is used here as the `toArray` method of the PriorityQueue does not return a necessarily sorted array.

### CODE REVIEW

File: core/src/test/scala/org/apache/spark/rdd/PairRDDFunctionsSuite.scala [hunk 1]
```diff
@@ -47,6 +47,19 @@ class PairRDDFunctionsSuite extends FunSuite with SharedSparkContext {
     assert(valuesFor5.toList.sorted === List(1, 3))
   }
 
+  test("topByKey") {
+    val pairs = sc.parallelize(Array((1, 1), (1, 2), (3, 2), (3, 7), (3, 5), (5, 1), (5, 3)), 2)
+
+    val sets = pairs.topByKey(2).collect()
```
Comment (mengxr) [inline core/src/test/scala/org/apache/spark/rdd/PairRDDFunctionsSuite.scala]: Use `colelctAsMap`.

File: core/src/main/scala/org/apache/spark/rdd/PairRDDFunctions.scala [hunk 2]
```diff
@@ -163,6 +163,38 @@ class PairRDDFunctions[K, V](self: RDD[(K, V)])
...
+
+  /**
+   * Returns the first k (smallest) elements for each key from this RDD as defined by the specified
+   * implicit Ordering[T] and maintains the ordering. This does the opposite of [[topByKey]].
+   *
+   * @param num k, the number of elements to return
+   * @param ord the implicit ordering for T
+   * @return an RDD that contains the smallest k values for each key
+   */
+  def takeOrderedByKey(num: Int)(implicit ord: Ordering[V]): RDD[(K, Array[V])] = {
```
Comment (mengxr) [inline core/src/main/scala/org/apache/spark/rdd/PairRDDFunctions.scala]: I think having `topByKey` should be sufficient. Users can reverse the order easily. @rxin

File: core/src/main/scala/org/apache/spark/rdd/PairRDDFunctions.scala [hunk 3]
```diff
@@ -163,6 +163,38 @@ class PairRDDFunctions[K, V](self: RDD[(K, V)])
   }
 
   /**
+   * Returns the top k (largest) elements for each key from this RDD as defined by the specified
+   * implicit Ordering[T]. This does the opposite of [[takeOrderedByKey]].
```
Comment (mengxr) [inline core/src/main/scala/org/apache/spark/rdd/PairRDDFunctions.scala]: Please document the behavior when there are less than k elements in a group.

File: core/src/main/scala/org/apache/spark/rdd/PairRDDFunctions.scala [hunk 4]
```diff
@@ -163,6 +163,28 @@ class PairRDDFunctions[K, V](self: RDD[(K, V)])
...
+    aggregateByKey(new BoundedPriorityQueue[V](num)(ord))(
+      seqOp = (queue, item) => {
+        queue += item
+        queue
+      },
+      combOp = (queue1, queue2)  => {
+        queue1 ++= queue2
+        queue1
+      }
+    ).mapValues(_.toArray.sorted(ord.reverse))
```
Comment (srowen) [inline core/src/main/scala/org/apache/spark/rdd/PairRDDFunctions.scala]: Does the `toArray` not already give you the top k in order? this seems to be the behavior already as it returns the array formed from the iterator in the underlying `PriorityQueue`. Worth testing I think. (That said, I suppose sorting an already-sorted array is pretty fast.)

Nit: a few lines above, an extra space in front of `=>`

**Comment by srowen** [comment]
Looking good, but what about the Java / Python APIs too?

File: core/src/main/scala/org/apache/spark/rdd/PairRDDFunctions.scala (hunk 4 above)
Comment (coderxiang) [inline core/src/main/scala/org/apache/spark/rdd/PairRDDFunctions.scala]: Good catch. Removing `sorted` will fail the test as the `toArray` does not generate a sorted sequence.

**Comment by rxin** [comment]
Thanks for the contribution.

I think we should first put this in MLlib, and then consider moving it to RDD in the future. The RDD API is fairly crowded, and adding more functions to it will make it harder to learn/read/understand.

**Comment by coderxiang** [comment]
@rxin @mengxr per the comments, I created `MLPairRDDFunctions.scala` and moved the function there in the update.

File: core/src/main/scala/org/apache/spark/rdd/PairRDDFunctions.scala (hunk 4 above)
Comment (srowen) [inline core/src/main/scala/org/apache/spark/rdd/PairRDDFunctions.scala]: Hm OK that surprises me but if you verified it is required, leave it of course

File: mllib/src/test/scala/org/apache/spark/mllib/rdd/MLPairRDDFunctionsSuite.scala [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+import org.apache.spark.mllib.rdd.MLPairRDDFunctions._
+
+class MLPairRDDFunctionsSuite extends FunSuite with MLlibTestSparkContext {
+  test("topByKey") {
+    val pairs = sc.parallelize(Array((1, 1), (1, 2), (3, 2), (3, 7), (3, 5), (5, 1), (5, 3)), 2)
+
+    val topMap = pairs.topByKey(2).collectAsMap()
+
+    assert(topMap.size === 3)
+    val valuesFor1 = topMap(1)
```
Comment (mengxr) [inline mllib/src/test/scala/org/apache/spark/mllib/rdd/MLPairRDDFunctionsSuite.scala]: We don't need this variable. Just use `topMap(1)` in assert.

**Comment by mengxr** [comment]
LGTM. Merged into master. Thanks!

**Comment by coderxiang** [comment]
Thanks!

**Pull request closed without merging**